src/Traits/HasRoles.php

```diff
@@ -84,6 +84,59 @@ public function scopeRole(Builder $query, $roles): Builder
     }
 
     /**
+     * Scope the model query to certain permissions only.
+     *
+     * @param \Illuminate\Database\Eloquent\Builder $query
+     * @param string|array|\Spatie\Permission\Contracts\Permission|\Illuminate\Support\Collection $permissions
+     *
+     * @return \Illuminate\Database\Eloquent\Builder
+     */
+    public function scopePermission(Builder $query, $permissions): Builder
+    {
+        if ($permissions instanceof Collection) {
+            $permissions = $permissions->toArray();
+        }
+
+        if (! is_array($permissions)) {
+            $permissions = [$permissions];
+        }
+
+        $rolesWithPermissions = collect([]);
+        $permissions = array_map(function ($permission) use (&$rolesWithPermissions) {
+            if ($permission instanceof Permission) {
+                $permissionModel = $permission;
+            } else {
```

Comment:
O how I dislike `else` statements. Refactor this to it's own `protected` function where you can use an early return.